### PR TITLE
Devops 157 support developers

### PR DESCRIPTION
**Pull request by deepfire**
- dev-env key generation
- DNS provisioning in the `aws.iohkdev.io` domain

### CODE REVIEW

File: iohk/iohk-ops.hs [hunk 1]
```diff
@@ -289,4 +289,44 @@ runTemplate o@Options{..} Template{..} args = do
...
+      sh $ do
+        k <- Turtle.find ((prefix "keys/keys-testnet/rich/key") <> (suffix ".sk"))
+          "keys/keys-testnet/rich"
+        cp k $ "keys" </> Path.filename k
+  echo "Cluster deployment has been prepared."
+
+runNew _ _ _ = error "impossible"
+
+-- | Use 'cardano-keygen' to create keys for a develoment cluster.
+generateDevKeys :: Options -> ConfigurationKey -> Turtle.FilePath -> IO ()
```
Comment (domenkozar) [inline iohk/iohk-ops.hs]: This should be named `generateGenesis` as a general case, then used for development clusters as needed.

File: modules/development.nix [hunk 2]
```diff
@@ -7,8 +7,8 @@ with import ./../lib.nix;
     global = {
       allocateElasticIP            = false;
       enableEkgWeb                 = true;
-      dnsHostname                  = null;
-      dnsDomainname                = mkForce null;
+      dnsHostname                  = "${name}-${config.deployment.name}";
```
Comment (domenkozar) [inline modules/development.nix]: `name` has the same value as `config.deployment.name`, did you verify the resulting URL?

File: iohk/iohk-ops.hs (hunk 1 above)
Comment (deepfire) [reply iohk/iohk-ops.hs]: The genesis is generated by `cardano-node` during startup, so it's really just generation of keys compatible with a provided configuration key -- to the best of my understanding.

cc @gromakovsky

File: modules/development.nix (hunk 2 above)
Comment (deepfire) [reply modules/development.nix]: @domenkozar, are you sure?

I'm reading https://nixos.org/nixops/manual/#ssec-relnotes-1.3 differently.

Also: https://github.com/NixOS/nixops/blob/master/nix/eval-machine-info.nix#L73

Comment (deepfire) [reply modules/development.nix]: As to the DNS provisioning -- I've tested it, indeed.

Comment (domenkozar) [reply modules/development.nix]: You're right, `name` is machine name.

File: iohk/iohk-ops.hs (hunk 1 above)
Comment (domenkozar) [reply iohk/iohk-ops.hs]: Why is there parameter called `--genesis-out-dir`?

In that case we should give it a bit descriptive name, like `generateStakeKeys`.

Comment (gromakovsky) [reply iohk/iohk-ops.hs]: Probably because @pva701 is bad at naming things :)
In master we also have `dump-genesis-data --path PATH` command.
I can rename `generate-keys-by-spec` to `dump-genesis-secrets` (not keys, because it also dumps AVVM seeds) and its path argument to `--path`. I'd like to do it in master, because it's not something strictly necessary for 1.0.

**Comment by deepfire** [comment]
@domenkozar, CI is green now.

**Comment by domenkozar** [comment]
@deepfire is it complete so I can re-review? Note that it has git conflicts.

**Comment by deepfire** [comment]
@domenkozar, fixed the `cardano-sl-src.json` bumping conflict.

Answering your question -- I believe I have addressed all of your review comments, yes.

File: iohk/NixOps.hs [hunk 3]
```diff
@@ -820,29 +519,33 @@ cmd   Options{..} bin args = do
...

 -- * Invoking nixops
 --
-iohkNixopsPath :: FilePath -> FilePath
-iohkNixopsPath defaultNix =
-  let storePath  = Sys.unsafePerformIO $ inprocs "nix-build" ["-A", "nixops", format fp defaultNix] $
+iohkNixopsPath :: FilePath
+iohkNixopsPath =
+  let defaultNix = "default.nix"
+      storePath  = Sys.unsafePerformIO $ inprocs "nix-build" ["-A", "nixops", format fp defaultNix] $
```
Comment (domenkozar) [inline iohk/NixOps.hs]: Why not just make it an `IO` function?

File: iohk/NixOps.hs [hunk 4]
```diff
@@ -820,29 +519,33 @@ cmd   Options{..} bin args = do
...
-  executor o (format fp $ iohkNixopsPath "default.nix")
-  (fromCmd com : nixopsCmdOptions o c <> fmap fromArg args)
+  executor o (format fp iohkNixopsPath)
+    (fromCmd com : nixopsCmdOptions o c <> fmap fromArg args)
 
-nixops' :: Options -> NixopsConfig -> NixopsCmd -> [Arg] -> IO (ExitCode, Text)
-nixops  :: Options -> NixopsConfig -> NixopsCmd -> [Arg] -> IO ()
-nixops' = nixops'' cmd'
-nixops  = nixops'' cmd
+nixops'  :: Options -> NixopsConfig -> NixopsCmd -> [Arg] -> IO (ExitCode, Text)
```
Comment (domenkozar) [inline iohk/NixOps.hs]: It's really really hard to follow functions like this. There are 5 functions used in below two lines with almost indentical names. Why not give them appropriate names?

File: iohk/iohk-ops.hs (hunk 1 above)
Comment (domenkozar) [reply iohk/iohk-ops.hs]: I'd still call this function `generateStakeKeys` as that's what it does, the fact that it's developer environment comes from it's usage, not definition.

File: iohk/NixOps.hs (hunk 3 above)
Comment (deepfire) [reply iohk/NixOps.hs]: The intent here is to calculate it just once -- an IO action would be recomputed repeatedly.

File: iohk/NixOps.hs (hunk 4 above)
Comment (deepfire) [reply iohk/NixOps.hs]: I believe I didn't change this code in this PR.  Can we address this later?

Comment (deepfire) [reply iohk/NixOps.hs]: Also, do you have any suggestions?

File: iohk/NixOps.hs [hunk 5]
```diff
@@ -820,29 +519,33 @@ cmd   Options{..} bin args = do
...
-iohkNixopsPath defaultNix =
-  let storePath  = Sys.unsafePerformIO $ inprocs "nix-build" ["-A", "nixops", format fp defaultNix] $
+iohkNixopsPath :: FilePath
+iohkNixopsPath =
+  let defaultNix = "default.nix"
+      storePath  = Sys.unsafePerformIO $ inprocs "nix-build" ["-A", "nixops", format fp defaultNix] $
                    (trace (T.unpack $ format ("INFO: using "%fp%" expression for its definition of 'nixops'") defaultNix) empty)
-      nixopsPath = Path.fromText $ T.strip storePath <> "/bin/nixops"
-  in trace (T.unpack $ format ("INFO: nixops is "%fp) nixopsPath) nixopsPath
+      opsPath    = Path.fromText $ T.strip storePath <> "/bin/nixops"
```
Comment (domenkozar) [inline iohk/NixOps.hs]: We have `nixops` and `iohk-ops`, so `ops` is an overloaded term, why make the name binding ambigious?

File: iohk/NixOps.hs (hunk 4 above)
Comment (domenkozar) [reply iohk/NixOps.hs]: Sure

File: iohk/NixOps.hs (hunk 5 above)
Comment (deepfire) [reply iohk/NixOps.hs]: This is a local variable, and the meaning is very clear, IMO -- its usage is really spread across three lines.   But ok.
